internal/util/debug/debug.go

```diff
@@ -141,39 +174,55 @@ func RunHandler(ctx context.Context, addr string, r prometheus.Registerer, l *za
 		http.Redirect(rw, req, "/debug", http.StatusSeeOther)
 	})
 
+	lis, err := net.Listen("tcp", opts.TCPAddr)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	return &Handler{
+		opts:     opts,
+		lis:      lis,
+		handlers: handlers,
+		stdL:     stdL,
+	}, nil
+}
+
+// Serve runs debug handler until ctx is canceled.
+//
+// It exits when handler is stopped and listener closed.
+func (h *Handler) Serve(ctx context.Context) {
 	s := http.Server{
-		Addr:     addr,
-		ErrorLog: stdL,
+		Addr:     h.opts.TCPAddr,
+		ErrorLog: h.stdL,
 		BaseContext: func(_ net.Listener) context.Context {
 			return ctx
 		},
 	}
 
-	go func() {
-		lis := must.NotFail(net.Listen("tcp", addr))
+	root := fmt.Sprintf("http://%s", h.lis.Addr())
 
-		root := fmt.Sprintf("http://%s", lis.Addr())
+	h.opts.L.Sugar().Infof("Starting debug server on %s ...", root)
 
-		l.Sugar().Infof("Starting debug server on %s ...", root)
+	paths := maps.Keys(h.handlers)
+	slices.Sort(paths)
 
-		paths := maps.Keys(handlers)
-		slices.Sort(paths)
-
-		for _, path := range paths {
-			l.Sugar().Infof("%s%s - %s", root, path, handlers[path])
-		}
+	for _, path := range paths {
+		h.opts.L.Sugar().Infof("%s%s - %s", root, path, h.handlers[path])
+	}
 
-		if err := s.Serve(lis); !errors.Is(err, http.ErrServerClosed) {
+	go func() {
+		if err := s.Serve(h.lis); !errors.Is(err, http.ErrServerClosed) {
 			panic(err)
 		}
 	}()
 
 	<-ctx.Done()
 
-	stopCtx, stopCancel := context.WithTimeout(context.Background(), time.Second)
-	defer stopCancel()
-	s.Shutdown(stopCtx) //nolint:contextcheck // use new context for cancellation
+	stopCtx, stopCancel := ctxutil.WithDelay(ctx)
+	defer stopCancel(nil)
+	_ = s.Shutdown(stopCtx)
```

Comment:
By the way, why don't we log potential errors? There are no potentially valuable errors for us? E.g. if context's deadline is exceeded and the server was force-stopped, it might be an interesting learning.

Reply:
I don't think there is a way to make those functions return errors